### PR TITLE
[ENH] golden output tests for `LagLlamaForecaster`

### PR DESCRIPTION
**Pull request by geetu040**
Towards #10672

Add regression tests that validate lag-llama's sktime implementation outputs against outputs from the original/source implementation ([time-series-foundation-models/lag-llama](https://github.com/time-series-foundation-models/lag-llama)).

<details>

<summary>
code used to generate the golden outputs for each case
</summary>

```py

"""Generate reference outputs directly from the Lag-Llama source implementation.

This script mirrors exactly what sktime.forecasting.lagllama.LagLlamaForecaster
does internally (GluonTS dataset construction, estimator/predictor setup,
seeded sampling in ``make_evaluation_predictions``) but calls
``lag_llama.gluon.estimator.LagLlamaEstimator`` directly, without importing
sktime's forecaster.
"""

import random

import numpy as np
import pandas as pd
import torch
from gluonts.dataset.pandas import PandasDataset
from gluonts.evaluation import make_evaluation_predictions
from huggingface_hub import hf_hub_download
from lag_llama.gluon.estimator import LagLlamaEstimator


def load_airline_values():
    from sktime.datasets import load_airline

    y = load_airline()
    return y


def get_checkpoint_path():
    return hf_hub_download(
        repo_id="time-series-foundation-models/Lag-Llama",
        filename="lag-llama.ckpt",
    )


def build_dataset(y_train, prediction_length):
    # ``make_evaluation_predictions`` always treats the tail ``prediction_length``
    # points of each series as the (withheld) future and only feeds the rest as
    # context to the predictor. To forecast on the *real* context (not a
    # truncated one), placeholder rows must be appended for it to strip off,
    # mirroring LagLlamaForecaster._extend_df.
    df = y_train.to_frame(name="value").astype("float32")
    df.index.name = "timepoints"
    future_index = pd.period_range(
        df.index[-1] + 1, periods=prediction_length, freq=df.index.freq
    )
    future_df = pd.DataFrame(
        {"value": np.zeros(prediction_length, dtype="float32")}, index=future_index
    )
    ext_df = pd.concat([df, future_df])
    return PandasDataset(
        ext_df, future_length=0, target="value", freq=ext_df.index.freqstr
    )


def run_case(
    prediction_length=16,
    context_length=32,
    num_samples=100,
    batch_size=1,
    use_rope_scaling=False,
    nonnegative_pred_samples=False,
    device="cpu",
):
    device = torch.device(device)
    y = load_airline_values()
    y_train = y.iloc[:-prediction_length]
    dataset = build_dataset(y_train, prediction_length)

    ckpt_path = get_checkpoint_path()
    ckpt = torch.load(ckpt_path, map_location=device, weights_only=False)
    estimator_args = ckpt["hyper_parameters"]["model_kwargs"]

    rope_scaling_arguments = None
    if use_rope_scaling:
        rope_scaling_arguments = {
            "type": "linear",
            "factor": max(
                1.0,
                (context_length + prediction_length) / estimator_args["context_length"],
            ),
        }

    estimator = LagLlamaEstimator(
        ckpt_path=ckpt_path,
        prediction_length=prediction_length,
        context_length=context_length,
        input_size=estimator_args["input_size"],
        n_layer=estimator_args["n_layer"],
        n_embd_per_head=estimator_args["n_embd_per_head"],
        n_head=estimator_args["n_head"],
        scaling=estimator_args["scaling"],
        time_feat=estimator_args["time_feat"],
        rope_scaling=rope_scaling_arguments,
        batch_size=batch_size,
        num_parallel_samples=num_samples,
        device=device,
        nonnegative_pred_samples=nonnegative_pred_samples,
    )

    original_load = torch.load

    def patched_load(*args, **kwargs):
        kwargs["weights_only"] = False
        return original_load(*args, **kwargs)

    torch.load = patched_load
    try:
        lightning_module = estimator.create_lightning_module()
        transformation = estimator.create_transformation()
        predictor = estimator.create_predictor(transformation, lightning_module)
    finally:
        torch.load = original_load

    np_state = np.random.get_state()
    py_state = random.getstate()
    np.random.seed(0)
    random.seed(0)
    try:
        with torch.random.fork_rng():
            torch.manual_seed(0)
            forecast_it, _ = make_evaluation_predictions(
                dataset=dataset, predictor=predictor, num_samples=num_samples
            )
            forecasts = list(forecast_it)
    finally:
        np.random.set_state(np_state)
        random.setstate(py_state)

    point_forecast = forecasts[0].mean_ts.to_numpy()
    return point_forecast[:3]


if __name__ == "__main__":
    print("=== context_length=32, num_samples=100, no rope scaling ===")
    print(run_case(context_length=32, num_samples=100))

    print("=== context_length=64, num_samples=100, rope scaling ===")
    print(run_case(context_length=64, num_samples=100, use_rope_scaling=True))

    print("=== context_length=32, num_samples=20, nonnegative samples ===")
    print(run_case(context_length=32, num_samples=20, nonnegative_pred_samples=True))


```
</details>

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
The stochastic nature of the forecaster may cause sporadic failures.

Can you please:

* check whether setting a random seed is possible
* otherwise set the tolerance higher, so the error probability is less than 0.01%

**Comment by geetu040** [comment]
yes I think that is what's happening, I've updated the tolerance with an explanation in the comments

**Comment by fkiraly** [comment]
Are you certain? The rtol was 0.08 previously and still is.

With expected/mean deviation of 0.05, this seems risky.

Either way, I do not see a commit that actually updates the tolerance (as you claim above).

**Comment by geetu040** [comment]
The rtol was changed to `0.08` in afbef990cd24873d86098802c4c7e2a7c39dc842 after which the CI passes.

> With expected/mean deviation of 0.05, this seems risky.

I think `0.08` looked reasonable given that for these `num_samples` it should not deviate that much. I  have tried running it in multiple iterations.

**Comment by fkiraly** [comment]
I do not follow your calculation here - may I kindly request you do a repeat execution with a mean/std calculation to ensure we have a rough bound on the number of sporadic failures?

**Comment by geetu040** [comment]
> I do not follow your calculation here - may I kindly request you do a repeat execution with a mean/std calculation to ensure we have a rough bound on the number of sporadic failures?

@fkiraly I think it's fine now, this was rather a device problem, looked into the forked code and fixed it there, the standard rtol/atol should work now.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great!

What I meant though is that you repeat the execution multiple times, and see what the standard deviation and mean are. Thean you make the tolerance such that the test only fires if we are six times standard deviation away from the mean.

**Comment by geetu040** [comment]
> What I meant though is that you repeat the execution multiple times, and see what the standard deviation and mean are. Thean you make the tolerance such that the test only fires if we are six times standard deviation away from the mean.

yes I see that, you were asking for something similar to https://github.com/sktime/sktime/pull/10736#issuecomment-5244587823, but I just realized that the outputs in the CI were same for all the machines and different only locally which was caused because of a device problem in vendored code which is fixed now, the outputs are matching and therefore we don't need to repeat the execution to find the mean and std.

**Review by fkiraly — APPROVED** [review]
Makes sense.
